rust/src/dataset/transaction.rs

```diff
@@ -0,0 +1,555 @@
+                .unwrap_or(0)
+        };
+        let mut final_fragments = Vec::new();
+
+        let maybe_existing_fragments =
+            current_manifest
+                .map(|m| m.fragments.as_ref())
+                .ok_or_else(|| Error::Internal {
+                    message: format!(
+                        "No current manifest was provided while building manifest for operation {}",
+                        self.name()
+                    ),
+                });
+
+        match &self.operation {
+            Operation::Append { ref fragments } => {
+                final_fragments.extend(maybe_existing_fragments?.clone());
+                final_fragments.extend(Self::fragments_with_ids(
+                    fragments.clone(),
+                    &mut fragment_id,
+                ));
+            }
+            Operation::Delete {
+                ref updated_fragments,
+                ref deleted_fragment_ids,
+                ..
+            } => {
+                // Remove the deleted fragments
+                final_fragments.extend(maybe_existing_fragments?.clone());
+                final_fragments.retain(|f| !deleted_fragment_ids.contains(&f.id));
+                final_fragments.iter_mut().for_each(|f| {
+                    for updated in updated_fragments {
+                        if updated.id == f.id {
+                            *f = updated.clone();
+                        }
+                    }
+                });
+            }
+            Operation::Overwrite { ref fragments, .. } => {
+                final_fragments.extend(Self::fragments_with_ids(
+                    fragments.clone(),
+                    &mut fragment_id,
+                ));
+            }
+            Operation::Rewrite {
+                ref new_fragments, ..
+            } => {
+                final_fragments.extend(Self::fragments_with_ids(
+                    new_fragments.clone(),
+                    &mut fragment_id,
+                ));
+            }
+            Operation::CreateIndex { .. } => {
+                final_fragments.extend(maybe_existing_fragments?.clone())
+            }
+            Operation::Merge { ref fragments, .. } => {
+                final_fragments.extend(fragments.clone());
+            }
+        };
+
+        let mut manifest = if let Some(current_manifest) = current_manifest {
+            Manifest::new_from_previous(current_manifest, &schema, Arc::new(final_fragments))
+        } else {
+            Manifest::new(&schema, Arc::new(final_fragments))
+        };
+
+        manifest.tag = self.tag.clone();
+
+        if config.auto_set_feature_flags {
+            apply_feature_flags(&mut manifest);
+        }
+        manifest.set_timestamp(config.timestamp);
+
+        manifest.update_max_fragment_id();
+
+        manifest.transaction_file = Some(transaction_file_path.to_string());
+
+        Ok(manifest)
+    }
+}
+
+impl TryFrom<&pb::Transaction> for Transaction {
```

Comment:
Not a real comment.  Just bleh, rust protobuf seems to require quite a bit of boilerplate code.

Reply:
Ha. I think if we want to operate directly off of the generated message structs we could, it's just then we'd have to deal with all the optional fields throughout the code base. So I think we create these conversions to stricter structs just to keep the validation in one place. But it does involve some boilerplate.